query-engine/prisma-models/src/internal_data_model.rs

```diff
@@ -58,12 +58,28 @@ impl InternalDataModel {
             .ok_or_else(|| DomainError::ModelNotFound { name: name.to_string() })
     }
 
-    pub fn find_relation(&self, name: &str) -> crate::Result<RelationWeakRef> {
+    /// This method takes the two models at the ends of the relation as a first argument, because
+    /// relation names are scoped by the pair of models in the relation. Relation names are _not_
+    /// globally unique.
+    pub fn find_relation(&self, model_names: (&str, &str), relation_name: &str) -> crate::Result<RelationWeakRef> {
         self.relations
             .get()
-            .and_then(|relations| relations.iter().find(|relation| relation.name == name))
+            .and_then(|relations| {
+                // The scope for a relation name is only between two models. Every pair of models
+                // has its own scope for relation names.
+                relations.iter().find(|relation| {
+                    relation.name == relation_name
+                        && ((relation.is_self_relation()
```

Comment:
Wow that is a pretty incredible if statement. It might be nicer to put that into a small function or do you think its best to leave as is?

Reply:
This should be addressed in [`d3f7e98` (#3169)](https://github.com/prisma/prisma-engines/pull/3169/commits/d3f7e987f4ec4f4234cc0144d1d8a20e2150ecb1) — wdyt?

Reply:
So much easier to read and I really like the comment. Thanks